authz: update troubleshooting guide for authZ policy (#6767)
istio/istio.io: content/en/docs/ops/common-problems/security-issues/index.md

1. Make sure there is no typo in your policy YAML file.

1. Avoid enabling authorization for istiod. Istio authorization policy is designed for authorizing access to workloads in Istio Mesh. Enabling it for istiod may cause unexpected behavior.

1. Make sure that your authorization policies are in the right same namespace (by checking "metadata"/”namespace” line).

1. Make sure that your authorization policies don't use any HTTP only fields for TCP traffic. Otherwise, Istio ignores the policies as if they didn't exist.

> 

Reviewer: didn't exist =>don't exist. 

It that true for deny policies?

Author: Yeah, it's slightly different here. Updated and added another bullet point for deny policy.

Reviewer: Looks good. Thanks.

> 1. Make sure there is no typo in your policy YAML file.

Reviewer 2: This seems unhelpful. The yaml parser and analyzer should complain about many typos. I would suggest instead describing specific common typos that can slip through, or remove this item. Just saying "typo" is too general IMO.

Author: I think the yaml parser and the analyzer doesn't catch the typos, for example, if I apply the following policy:
```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
 name: test
 namespace: foo
spec:
 action: ALLOW
 selector:
   matchLabels:
     app: test
 typo_rules:
 - from:
   - source:
       principals: ["cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account"]
```

the `typo_rules` field will not be caught and the policy will be accepted. The unknown fields will just be ignored when converting to the AuthorizationPolicy protobuf in Pilot.

1. Avoid enabling authorization for istiod. Istio authorization policy is designed for authorizing access to workloads in Istio Mesh. Enabling it for istiod may cause unexpected behavior.

> 1. Make sure that your authorization policies are in the right namespace (by checking "metadata"/”namespace” line).

Reviewer 2: ```suggestion
1. Make sure that your authorization policies are in the right namespace (as specified in `metadata/namespace` field).
```

1. Make sure that your authorization policies with ALLOW action don't use any HTTP only fields for TCP traffic.
Otherwise, Istio ignores the ALLOW policies as if they don't exist.

> 1. Make sure that your authorization policies with DENY action don't use any HTTP only fields for TCP traffic.

Reviewer 2: This and the above entry seem to be a convoluted way of saying that the HTTP only fields are ignored, which is what it already says in https://deploy-preview-6767--preliminary-istio.netlify.com/docs/concepts/security/#using-istio-authorization-on-plain-tcp-protocols. I would suggest just one entry that says something like:
```
Make sure that authorization policies for TCP workloads don't use any HTTP only fields.
HTTP only fields are ignored for TCP traffic and consequently have no effect on the policy.
```

1. Use `Ctrl+C` in the terminal you started in step 1 to stop the port-forward command.

1. Print the log of Istiod and search for `rbac` with the following command:

    {{< tip >}}
    You probably need to first delete and then re-apply your authorization policies so that
    the debug output is generated for these policies.
    {{< /tip >}}

    {{< text bash >}}
    $ kubectl logs $(kubectl -n istio-system get pods -l app=istiod -o jsonpath='{.items[0].metadata.name}') -c discovery -n istio-system | grep rbac
    {{< /text >}}

1. Check the output and verify:

    - There are no errors.

>     - There is a `building v1beta1 policy` message which means the filter is generated

Reviewer 2: ```suggestion
    - There is a `building v1beta1 policy` message which indicates the filter was generated
```

    2020-03-05T23:43:21.623543Z   debug   rbac   building filter for TCP listener protocol
    2020-03-05T23:43:21.623546Z   debug   rbac   building v1beta1 policy
    2020-03-05T23:43:21.623572Z   debug   rbac   constructed internal model: &{Permissions:[{Services:[] Hosts:[] NotHosts:[] Paths:[] NotPaths:[] Methods:[] NotMethods:[] Ports:[] NotPorts:[] Constraints:[] AllowAll:true v1beta1:true}] Principals:[{Users:[] Names:[cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account] NotNames:[] Group: Groups:[] NotGroups:[] Namespaces:[] NotNamespaces:[] IPs:[] NotIPs:[] RequestPrincipals:[] NotRequestPrincipals:[] Properties:[] AllowAll:false v1beta1:true}]}
    2020-03-05T23:43:21.623625Z   debug   rbac   generated policy ns[foo]-policy[ext-authz-server]-rule[0]: permissions:<and_rules:<rules:<any:true > > > principals:<and_ids:<ids:<or_ids:<ids:<authenticated:<principal_name:<exact:"spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account" > > > > > > >
    2020-03-05T23:43:21.623645Z   debug   rbac   added TCP filter to filter chain 0
    2020-03-05T23:43:21.623648Z   debug   rbac   added TCP filter to filter chain 1
    {{< /text >}}

>     It means istiod generated:

Reviewer 2: ```suggestion
    This shows that Istiod generated:
```

>     - A HTTP filter config with policy `ns[foo]-policy[ext-authz-server]-rule[0]` for workload with labels `app=ext-authz-server,...`.

Reviewer 2: ```suggestion
    - An HTTP filter config with policy `ns[foo]-policy[ext-authz-server]-rule[0]` for workload with labels `app=ext-authz-server,...`.
```

    - A TCP filter config with policy `ns[foo]-policy[ext-authz-server]-rule[0]` for workload with labels `app=ext-authz-server,...`.

> ## Ensure Pilot distributes policies to proxies correctly

Reviewer: Pilot => Istiod?
